### DISCUSSION TITLE
<.Resources.ByType>: can't evaluate field Resources in type *hugolib.ShortcodeWithPage`

### DISCUSSION THREAD

**Discussion by cnichte** [gethyas/hyas#440] [General]
Hi, i am building a simple custom  shortcode with `Hyas` and `doks`.  But it fails with this very first statement:

`{{ $images := .Resources.ByType "image" }}`

There is nothing special about the line, and it is shown in dozens of examples and executed without errors. I suspect that it's because of how Hyas integrates Hugo? Could that be? 

```
execute of template failed: template: shortcodes/test.html:5:32: executing "shortcodes/test.html" at <.Resources.ByType>: can't evaluate field Resources in type *hugolib.ShortcodeWithPage`
```

I am grateful for any advice :-)

**Comment by h-enk** [comment]
> There is nothing special about the line, and it is shown in dozens of examples and executed without errors. I suspect that it's because of how Hyas integrates Hugo? Could that be?

LGTM/No/No — Doks (Hyas) uses [this same statement](https://github.com/gethyas/doks-core/blob/main/layouts/blog/single.html#L11) to get a blog post image, and that works perfectly fine (see e.g. https://images.gethyas.com/blog/example-post/)

Did you [add the image(s) as a page resource](https://images.gethyas.com/docs/guides/image-resources/#page-resource)?

**Reply by cnichte** [reply]
Yes, i tested it in a blog post. The images reside in an images subfolder in the package. I try to build a carousel from images in a specific subfolder, as you can see here:

https://static.carsten-nichte.de/blog/call-it-corona-ist-erschienen/

* A Test shortcode without `.Resources.ByType` works great.
* Images in page resource are shown.
* Shortcode with  `.Resources.ByType` fails:

![Bildschirmfoto 2023-12-11 um 12 42 59](https://github.com/gethyas/hyas/assets/96977775/b76e849e-69aa-475e-a541-3b00f4fe8a70)

I have really no idea what's going on there
.
.
.

---
Remarks:

I have a website with a focus on photography (currently Wordpress). There is a lot of content to move.  So far 600 pages and over 1000 pictures. The pages are moved so far, but I'm struggling with the images (I also have a few questions about @hyas/images ;-) ..so, it's a bigger project.

* Here I build up my content:  https://static.carsten-nichte.de
* The "old" page:  https://carsten-nichte.de/

I need three things to get the images under control:

1. A carousel (I try to build that right now), but fail at the very first step :-D
2. A lightbox. I'll try that soon https://www.npmjs.com/package/fslightbox 
3. and a Mansory layout, like the one you see here: https://carsten-nichte.de/binary-voids/idx/ 
    * I haven't found anything suitable yet. 

Perpaps that where great additions to Hyas? 

In addition: there are other little things, such as a video player, an audio player, which are able to play self-hosted content, and other stuff :-D

**Reply by h-enk** [reply]
It looks like a Hugo thing: the `ByType` method apparently (tested/confirmed) can't reach images in an `images` subfolder, although suggested in [Page Resources](https://gohugo.io/content-management/page-resources/)

Would it be an issue to add them on the same level as `index.html`? See also:

- https://github.com/h-enk/doks-440/tree/main/content/blog/example
- https://doks-440.netlify.app/blog/example-post/

**Reply by h-enk** [reply]
See also:

- [Video shortcode — Doks](https://getdoks.org/docs/guides/shortcodes/#video)
- [Slideshow — W3Schools](https://www.w3schools.com/howto/howto_js_slideshow.asp) + other links under **Images**
- [Carousel — Bootstrap](https://getbootstrap.com/docs/5.3/components/carousel/)
- [Masonry — Bootstrap](https://getbootstrap.com/docs/5.3/components/card/#masonry)

**Reply by cnichte** [reply]
I put everything in the root directory of the bundle. The result is the same. Your `img` shortcode finds the image (also in any subdirectories) but my `test` shortcode hits the wall at `.Resources.ByType`. 

I have now also tried it with `{{ $images := .Resources.Match "images/carousel/*" }}`, with the images put in that specific folder...

```bash
execute of template failed at <.Resources.Match>: can't evaluate field Resources in type *hugolib.ShortcodeWithPage 
```

It seems that it does not know the .Resources object at all. How can that be? I'm really at a complete loss at this point. Total fail at basic stuff :-D

**Reply by h-enk** [reply]
It looks like you're trying to add a partial in Markdown content. Note, that there's a difference between [shortcodes](https://gohugo.io/content-management/shortcodes/) and [partials](https://gohugo.io/templates/partials/). Shortcodes are placed in your Markdown content files and get rendered to HTML. Partials are more like components and are referenced in your HTML layouts.

See for inspiration/background [gallery-slider](https://github.com/gethugothemes/hugo-modules/tree/master/gallery-slider)

**Reply by cnichte** [reply]
Okay,  Ive setup a clean Hugo project, with one Blog entry, as described in the Hugo QuickStart manual, and the error remains. There is something fundamentally wrong (or misunderstood by myself). Ive opened a Question in Hugo forum. Thanks so far :-) I will look in the other things when this problem is solved.

**Reply by h-enk** [reply]
I've set up a quick example to help you get started — see [Gallery Shortcode](https://doks-440.netlify.app/blog/example-post/)

Repo: https://github.com/h-enk/doks-440

**Reply by cnichte** [reply]
Soooo, here was my issue: "Each template receives context (the dot   `.`Resources).  Context is an object with values/methods. The context varies by template type, page kind, content, etc". Shortcode templates receive the following context:

* https://gohugo.io/quick-reference/methods/#shortcode

Shortcode doesnt now .Resources but .Page and .Page knows Ressources, so...

```
{{ $images := .Page.Resources.Match "images/carousel/*" }}
```

instead of

```
{{ $images := .Resources.Match "images/carousel/*" }}
```

Easy like that.  Now it works like a charm :-D
